### PR TITLE
feat(kuma-cp) Fault Injection entity

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary
Add `FaultInjection` entity

### Full changelog

* Add `*.proto`
* Add CRD
* Add to API
* Add support to `kumactl`

### CODE REVIEW

File: api/mesh/v1alpha1/fault_injection.proto [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+package kuma.mesh.v1alpha1;
+
+option go_package = "v1alpha1";
+
+import "google/protobuf/duration.proto";
+import "mesh/v1alpha1/selector.proto";
+import "google/protobuf/wrappers.proto";
+
+message FaultInjection {
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: We need doc-comments in this file. Take a look at other proto definitions.

File: api/mesh/v1alpha1/fault_injection.pb.validate.go [hunk 2]
```diff
@@ -0,0 +1,501 @@
+// Code generated by protoc-gen-validate. DO NOT EDIT.
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.pb.validate.go]: I don't think we need this.

File: app/kumactl/cmd/install/install_control_plane_test.go [hunk 3]
```diff
@@ -93,7 +93,7 @@ var _ = Describe("kumactl install control-plane", func() {
 			// then
 			Expect(actual).To(MatchYAML(expected))
 			// and
-			actualManifests := data.SplitYAML(data.File{Data: expected})
+			actualManifests := data.SplitYAML(data.File{Data: actual})
```
Comment (jakubdyszkiewicz) [inline app/kumactl/cmd/install/install_control_plane_test.go]: Good catch! I was having problems with this last time I edited the install script and was wondering what is going on.

**Comment by jakubdyszkiewicz** [comment]
don't forget about Kuma GUI but this comes with a separate PR. You can ask @bloqhead for help if needed.

**Comment by subnetmarco** [comment]
Please remember to also write documentation for this, and update the `CHANGELOG.md`.

File: api/mesh/v1alpha1/fault_injection.proto [hunk 4]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
 message FaultInjection {
 
+  // List of selectors to match dataplanes that are sources of traffic.
   repeated Selector sources = 1;
 
+  // List of selectors to match services that are destinations of traffic.
   repeated Selector destinations = 2;
 
+  // Configuration defines several types of faults, at least one delay should be
+  // specified
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: I think this should be rewritten to `Conf defines several types of faults, at least one fault should be specified`

File: api/mesh/v1alpha1/fault_injection.proto [hunk 5]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
   message Conf {
 
+    // Defines configuration of the delay
     message Delay {
+      // Percentage of requests on which delay will be injected
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: Also specify what are available values: [0.0-100.0] range

File: api/mesh/v1alpha1/fault_injection.proto [hunk 6]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
   repeated Selector sources = 1;
 
+  // List of selectors to match services that are destinations of traffic.
   repeated Selector destinations = 2;
 
+  // Configuration defines several types of faults, at least one delay should be
+  // specified
   message Conf {
 
+    // Defines configuration of the delay
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: `Delay defines configuration of delaying a response from a destination`

File: api/mesh/v1alpha1/fault_injection.proto [hunk 7]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
 import "mesh/v1alpha1/selector.proto";
 import "google/protobuf/wrappers.proto";
 
+// FaultInjection defines configuration of faults between dataplanes.
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: `the configuration`

I try to use Grammarly to catch articles and more.

File: api/mesh/v1alpha1/fault_injection.proto [hunk 8]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
 
+  // Configuration defines several types of faults, at least one delay should be
+  // specified
   message Conf {
 
+    // Defines configuration of the delay
     message Delay {
+      // Percentage of requests on which delay will be injected
       google.protobuf.DoubleValue percentage = 1;
+      // Duration, the value of delay
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: `The duration for which the response will be delayed`?
or something similar, `value of delay` could be more descriptive.

File: api/mesh/v1alpha1/fault_injection.proto [hunk 9]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
       google.protobuf.Duration value = 2;
     }
+    // Delay if specified makes source side to experience delay during requests
+    // to the destination
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: experience in which way? I think it's better to say that the response from the server will be delivered with the delay

File: api/mesh/v1alpha1/fault_injection.proto [hunk 10]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
     Delay delay = 1;
 
+    // Defines configuration of the abort
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: How about `Abort defines a configuration of not delivering requests to destination service and replacing the responses from destination dataplane by predefined status code`?

By the way - Is a request with to destination service delivered to the service and response is replaced or is it not delivered and Envoy on the destination side just responds with status code?

File: api/mesh/v1alpha1/fault_injection.proto [hunk 11]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
     message Abort {
+      // Percentage of requests on which abort will be injected
       google.protobuf.DoubleValue percentage = 1;
+      // HttpStatus code which will be returned to source side
       google.protobuf.UInt32Value httpStatus = 2;
     }
+    // Abort if specified makes source side to receive specified httpStatus code
     Abort abort = 2;
 
+    // Defines configuration of the response bandwidth
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: How about `ResponseBandwidth defines a configuration to limit the speed of responding to the requests`?

File: api/mesh/v1alpha1/fault_injection.proto [hunk 12]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
     message ResponseBandwidth {
+      // Percentage of requests on which response bandwidth limit will be
+      // injected
       google.protobuf.DoubleValue percentage = 1;
+      // Limit is represented by value measure in gbps, mbps, kbps or bps
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: Add an example. Is it `10kbps` or `10 kbps`?

File: api/mesh/v1alpha1/fault_injection.proto [hunk 13]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
       google.protobuf.StringValue limit = 2;
     }
+    // ResponseBandwidth if specified limits the speed of sending response body
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: response body or just response (headers + body etc.)?

File: api/mesh/v1alpha1/fault_injection.proto [hunk 14]
```diff
@@ -8,32 +8,52 @@ import "google/protobuf/duration.proto";
...
     }
+    // Delay if specified makes source side to experience delay during requests
+    // to the destination
     Delay delay = 1;
 
+    // Defines configuration of the abort
     message Abort {
+      // Percentage of requests on which abort will be injected
       google.protobuf.DoubleValue percentage = 1;
+      // HttpStatus code which will be returned to source side
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: `HTTP status code`

File: api/mesh/v1alpha1/fault_injection.proto (hunk 12 above)
Comment (lobkovilya) [reply api/mesh/v1alpha1/fault_injection.proto]: Both are acceptable now, I use the following regex for validation - `\d*\s?[gmk]?bps`. Do we want to leave one acceptable way to specify limit?

File: api/mesh/v1alpha1/fault_injection.proto (hunk 10 above)
Comment (lobkovilya) [reply api/mesh/v1alpha1/fault_injection.proto]: From envoy documentation: 
_If specified, the filter will **abort requests** based on the values in the object. At least abort or delay must be specified._
So I guess aborting requests means "not delivering them to service".

Comment (jakubdyszkiewicz) [reply api/mesh/v1alpha1/fault_injection.proto]: Yeah, it makes more sense for requests to be aborted on Envoy's side. Just wanted to make sure.

File: api/mesh/v1alpha1/fault_injection.proto (hunk 12 above)
Comment (jakubdyszkiewicz) [reply api/mesh/v1alpha1/fault_injection.proto]: Great, leave both. Just provide example of `10kbps`.

File: api/mesh/v1alpha1/fault_injection.proto (hunk 13 above)
Comment (lobkovilya) [reply api/mesh/v1alpha1/fault_injection.proto]: According to envoy doc "response body":
_The response rate limit to be applied to the response body of the stream. When configured, the percentage can be overridden by the fault.http.rate_limit.response_percent runtime key._

File: api/mesh/v1alpha1/fault_injection.proto [hunk 15]
```diff
@@ -17,37 +17,43 @@ message FaultInjection {
...
+      // [0.0 - 100.0] range
       google.protobuf.DoubleValue percentage = 1;
-      // Duration, the value of delay
+      // The duration during which the response will be delayed
       google.protobuf.Duration value = 2;
     }
-    // Delay if specified makes source side to experience delay during requests
-    // to the destination
+    // Delay if specified then response from the destination will be delivered
+    // with delay
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/fault_injection.proto]: a delay

File: pkg/core/resources/apis/mesh/fault_injection_validator.go [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+	faultResponseBandwidth := f.Spec.Conf.GetResponseBandwidth()
+	return faultResponseBandwidth != nil && !reflect.DeepEqual(*faultResponseBandwidth, v1alpha1.FaultInjection_Conf_ResponseBandwidth{})
+}
+
+func (f *FaultInjectionResource) validateSources() validators.ValidationError {
+	return ValidateSelectors(validators.RootedAt("sources"), f.Spec.GetSources(), ValidateSelectorsOpts{
+		RequireAtLeastOneSelector: true,
+		ValidateSelectorOpts: ValidateSelectorOpts{
+			RequireAtLeastOneTag: true,
+			RequireService:       true,
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/apis/mesh/fault_injection_validator.go]: I don't think service should be required in this entity.

File: pkg/core/resources/apis/mesh/fault_injection_validator.go [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+	faultAbort := f.Spec.Conf.GetAbort()
+	return faultAbort != nil && !reflect.DeepEqual(*faultAbort, v1alpha1.FaultInjection_Conf_Abort{})
+}
+
+func (f *FaultInjectionResource) HasFaultResponseBandwidth() bool {
+	faultResponseBandwidth := f.Spec.Conf.GetResponseBandwidth()
+	return faultResponseBandwidth != nil && !reflect.DeepEqual(*faultResponseBandwidth, v1alpha1.FaultInjection_Conf_ResponseBandwidth{})
+}
+
+func (f *FaultInjectionResource) validateSources() validators.ValidationError {
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/apis/mesh/fault_injection_validator.go]: Marco wanted to add `protocol: http` requirement to both sources and destinations

File: pkg/core/resources/apis/mesh/fault_injection_validator.go [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+		},
+	})
+}
+
+func (f *FaultInjectionResource) validateDestinations() validators.ValidationError {
+	return ValidateSelectors(validators.RootedAt("destinations"), f.Spec.GetDestinations(), ValidateSelectorsOpts{
+		RequireAtLeastOneSelector: true,
+		ValidateSelectorOpts: ValidateSelectorOpts{
+			RequireAtLeastOneTag: true,
+			RequireService:       true,
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/apis/mesh/fault_injection_validator.go]: I don't think service should be required in this entity.

File: pkg/core/resources/apis/mesh/fault_injection_validator.go (hunk 17 above)
Comment (lobkovilya) [reply pkg/core/resources/apis/mesh/fault_injection_validator.go]: Added additional validator, please take a look

**Pull request merged**